scio-tensorflow/src/main/scala/com/spotify/scio/tensorflow/TensorFlowDoFns.scala

```diff
@@ -18,31 +18,34 @@
 package com.spotify.scio.tensorflow
 
 import java.time.Duration
-import java.util.concurrent.{ConcurrentHashMap, ConcurrentMap}
-import java.util.function.Function
 
-import com.spotify.scio.transforms.DoFnWithResource
-import com.spotify.scio.transforms.DoFnWithResource.ResourceType
-import com.spotify.zoltar.tf.TensorFlowModel
-import com.spotify.zoltar.{Model, Models}
+import com.spotify.zoltar.tf.{TensorFlowLoader, TensorFlowModel}
+import com.spotify.zoltar.Model
 import org.apache.beam.sdk.transforms.DoFn
 import org.apache.beam.sdk.transforms.DoFn.{ProcessElement, Teardown}
-import org.tensorflow._
 import org.slf4j.LoggerFactory
-
+import org.tensorflow._
+import org.tensorflow.example.Example
 import scala.collection.JavaConverters._
 
-private[this] abstract class PredictDoFn[T, V, M <: Model[_]](
-  fetchOp: Seq[String],
-  inFn: T => Map[String, Tensor[_]],
-  outFn: (T, Map[String, Tensor[_]]) => V
-) extends DoFnWithResource[T, V, ConcurrentMap[String, M]] {
+import com.spotify.scio.transforms.DoFnWithResource
+import com.spotify.scio.transforms.DoFnWithResource.ResourceType
+import com.spotify.zoltar.Model.Id
+
+sealed trait PredictDoFn[T, V, M <: Model[_]] extends DoFnWithResource[T, V, M] {
   @transient private lazy val log = LoggerFactory.getLogger(this.getClass)
 
   def withRunner(f: Session#Runner => V): V
 
-  override def getResourceType: DoFnWithResource.ResourceType =
-    ResourceType.PER_CLASS
+  def extractInput(input: T): Map[String, Tensor[_]]
+
+  def extractOutput(input: T, out: Map[String, Tensor[_]]): V
+
+  def outputTensorNames: Seq[String]
+
+  override def createResource(): M
+
+  override def getResourceType: DoFnWithResource.ResourceType = ResourceType.PER_INSTANCE
```

Comment:
Had to make this run per instance, as we don't have control over resource identification
Without this, whichever model is loaded first in the JVM is always reused for the same class -- so a new instance with a different signature-def won't be loaded

Reply:
hmm, ok, I think this might actually be an issue I introduced in my changes. Before the resource was a concurrent map of models `Map<String, Model>`, I believe this solves (1) and (2) ]

Reply:
Yes, it would. I'm still curious if the multiple functions spawn on the same worker are clones or not. The resource method specifies 3 options, and talks about clone functions. So I'm curious as to when they would spawn up. If it works the way I suspect it does, then loading per instance shouldn't be an issue

Reply:
@regadas do you know about this?

Reply:
bumping this up @regadas ^^

Reply:
This is bad performance wise ... Why can't we keep the `Map<ModelId, Model>` as resource? `ModelId` can contain be based on the model signture def.

Regarding the `DoFn` `TearDown` to be honest is a bit unnecessary in this case; Models are never hot swapped or anything like that. They will be pretty much closed when pipeline is shutting down or being updated.

Reply:
I revered the resource to map. I'm not sure how clean-up will be dealt with though, since the model is loaded in tensorflow -- I guess if the machine/container gets teared down that cleanup might happen automagically, but it might leave behind some garbage/memory leaks

Reply:
The container will GC'ed so it should be ok. I'm actually getting an idea how to be a good citizen here ... but we can explore it in another PR.